ddp-workspace/projects/study-builder-ui/src/app/components/text-question-block/text-question-block.component.ts

```diff
@@ -13,35 +14,83 @@ import { SimpleTemplate } from '../../model/core-extended/simpleTemplate';
     angularClientBlock$: Observable<ActivityTextQuestionBlock>;
+    validationErrorMessages: string[] = [];
+    private blockCurrentAnswer = '';
+    private sub = new Subscription();
 
-    constructor(private config: ConfigurationService) {
-    }
-
+    constructor(
+        private config: ConfigurationService,
+        private questionConverter: ActivityQuestionConverter
+    ) {}
 
     ngOnInit(): void {
         this.angularClientBlock$ = this.definitionBlock$.pipe(
             tap(defBlock => console.log('getting defblock: %o', defBlock)),
             filter(block => !!block),
-            map(defBlock => this.buildFromDef(defBlock)));
+            map(defBlock => this.buildFromDef(defBlock)),
+            tap((block: ActivityTextQuestionBlock) => this.validate(block))
+        );
+    }
+
+    ngOnDestroy(): void {
+        this.sub.unsubscribe();
+    }
+
+    valueChanged(value: string): void {
+        this.blockCurrentAnswer = value;
+        const valueChangedSub = this.angularClientBlock$.subscribe((block: ActivityTextQuestionBlock) => this.validate(block, value));
+        this.sub.add(valueChangedSub);
     }
 
     private buildFromDef(defBlock: QuestionBlockDef<TextQuestionDef>): ActivityTextQuestionBlock {
-        const newClientBlock = new ActivityTextQuestionBlock();
         const questionDef = defBlock.question;
-        newClientBlock.inputType = questionDef.inputType;
-        newClientBlock.isRequired = questionDef.validations.some(val => val.ruleType === 'REQUIRED');
-        newClientBlock.stableId = questionDef.stableId;
+        const modifiedQuestionJson = {
+            ...questionDef,
+            validations: questionDef.validations.map(
+                // rename `ruleType` keys to `rule`
+                ({ruleType: rule, ...rest}) => ({rule, ...rest})
+            )
+        };
+
+        const newClientBlock = this.questionConverter.buildQuestionBlock(modifiedQuestionJson, null) as ActivityTextQuestionBlock;
+
         newClientBlock.placeholder = new SimpleTemplate(questionDef.placeholderTemplate)
             .getTranslationText(this.config.defaultLanguageCode);
         newClientBlock.question = new SimpleTemplate(questionDef.promptTemplate).getTranslationText(this.config.defaultLanguageCode);
+
+        if (this.blockCurrentAnswer) {
+            newClientBlock.setAnswer(this.blockCurrentAnswer, false);
+        }
         return newClientBlock;
     }
 
+    private validate(block: ActivityTextQuestionBlock, answer: string = block.answer): void {
+        block.setAnswer(answer || '', false);
+        for (const validator of block.validators) {
+            validator.recalculate();
+        }
+        this.validationErrorMessages = block.validators.map(validator => {
+            const result = validator.result;
+            return ActivityQuestionComponent.isActivityValidationResult(result) ? result.message : result;
+        });
+    }
+
```

Comment:
Could we find a way to refactor the execution of validators so that it can be shared/be the same for all the different types of questions?
Validation processing here looks pretty generic. Can we have this functionality to be shared with picklist component?

Reply:
Refactored - extracted base abstract component (6d13107018e52a92640197b9b630010313cbbab1, 0f11b98f03734431092d46589522bd0ca656ae74)